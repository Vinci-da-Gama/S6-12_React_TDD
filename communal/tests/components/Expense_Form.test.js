import React from 'react';
import moment from 'moment';
import { shallow } from 'enzyme';

import ExpenseForm from '../../src/components/livingExpense/Expense_Form';
import fixtureExpenses from '../fixtures/expenses.fixture';

let wrapper, addUrl = '/create', amountInput, singleDatePicker;

beforeEach(() => {
    wrapper = shallow(<ExpenseForm currentUrl={addUrl} />);
    amountInput = wrapper.find('input').at(1);
    singleDatePicker = wrapper.find('#sd-pIcker');
});

test('should render expense form correctly.', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should render expense form correctly with one expense data', () => {
    const wrapper = shallow(<ExpenseForm expense={fixtureExpenses[1]} currentUrl={addUrl} />);
    expect(wrapper).toMatchSnapshot();
});

test('should set new description.', () => {
    const value = 'This is new description.';
    wrapper.find('input').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('description')).toBe(value);
});

test('should set new amount.', () => {
    const value = '10.00';
    amountInput.simulate('change', {
        target: { value }
    });
    expect(wrapper.state('amount')).toBe(value);
});

test('should not set new amount, due to invalid value', () => {
    const value = '10.000';
    amountInput.simulate('change', {
        target: { value }
    });
    expect(wrapper.state('amount')).toBe('');
});

test('should set new date onDateChange.', () => {
    const now = moment();
    singleDatePicker.prop('onDateChange')(now);
    expect(wrapper.state('createdAt')).toEqual(now);
});

test('should set focus when onFocusChange.', () => {
    const focused = true;
    singleDatePicker.prop('onFocusChange')({ focused });
    expect(wrapper.state('calendarFocused')).toBe(focused);
});

test('should set new note.', () => {
    const value = 'This is new note.';
    wrapper.find('textarea[name="ef_note"]').simulate('change', {
        target: { value }
    });
    expect(wrapper.state('note')).toBe(value);
});

describe('Test form submit.', () => {

    test('should render error message for invaild form submission.', () => {
        wrapper.find('form[name="expenseForm"]').simulate('submit', {
            preventDefault: () => {}
        });
        expect(wrapper.state('error').length).toBeGreaterThan(0);
        expect(wrapper.state('error')).toBe('Please provide description and amount.');
        expect(wrapper).toMatchSnapshot();
    });

    test('should submit form successfully.', () => {
        const onSubmitSpy = jest.fn();
        const e0 = fixtureExpenses[0];
        const formWrapper = shallow(<ExpenseForm expense={e0} currentUrl={addUrl} onSubmit={onSubmitSpy} />);
        formWrapper.find('form[name="expenseForm"]').simulate('submit', {
            preventDefault: () => {}
        });
        expect(formWrapper.state('error')).toBe('');
        expect(onSubmitSpy).toHaveBeenLastCalledWith({
          description: e0.description,
          amount: e0.amount,
          note: e0.note,
          createdAt: e0.createdAt
        });
    });

});
