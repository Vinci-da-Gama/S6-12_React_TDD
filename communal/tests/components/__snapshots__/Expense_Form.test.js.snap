// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test form submit. should render error message for invaild form submission. 1`] = `
<div>
  <Alert
    closeAriaLabel="Close"
    color="danger"
    fade={true}
    isOpen={true}
    tag="div"
    transition={
      Object {
        "appear": true,
        "baseClass": "fade",
        "baseClassActive": "show",
        "enter": true,
        "exit": true,
        "in": true,
        "mountOnEnter": false,
        "onEnter": [Function],
        "onEntered": [Function],
        "onEntering": [Function],
        "onExit": [Function],
        "onExited": [Function],
        "onExiting": [Function],
        "tag": "div",
        "timeout": 150,
        "unmountOnExit": true,
      }
    }
  >
    Please provide description and amount.
  </Alert>
  <form
    name="expenseForm"
    noValidate={true}
    onSubmit={[Function]}
  >
    <Col
      className="mx-auto"
      md={4}
      sm={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Label
        className="bg-warning"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Description:
      </Label>
      <input
        autoFocus={true}
        className="w-100 form-control rounded-0"
        name="ef_description"
        onChange={[Function]}
        placeholder="Description..."
        type="text"
        value=""
      />
    </Col>
    <Col
      className="mx-auto"
      md={4}
      sm={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Label
        className="bg-warning"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Amount:
      </Label>
      <input
        className="w-100 form-control rounded-0"
        name="ef_Amount"
        onChange={[Function]}
        placeholder="Amount..."
        type="text"
        value=""
      />
    </Col>
    <Col
      className="mx-auto"
      md={4}
      sm={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Badge
        color="warning"
        pill={true}
        tag="span"
      >
        Create Date:
      </Badge>
      <br />
      <withStyles(SingleDatePicker)
        anchorDirection="left"
        appendToBody={false}
        block={false}
        calendarInfoPosition="bottom"
        customCloseIcon={null}
        customInputIcon={null}
        date={"1970-01-01T00:00:00.000Z"}
        daySize={39}
        disableScroll={false}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        horizontalMonthPadding={13}
        id="sd-pIcker"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepFocusOnInput={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        noBorder={false}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsSelected": [Function],
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month.",
            "jumpToPrevMonth": "Move backward to switch to the previous month.",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
            "moveFocusByOneMonth": "Switch months.",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
            "openThisPanel": "Open this panel.",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field.",
            "selectFocusedDate": "Select the date in focus.",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        regular={false}
        renderCalendarInfo={null}
        renderDayContents={null}
        renderMonthElement={null}
        renderMonthText={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        small={false}
        verticalHeight={null}
        verticalSpacing={22}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </Col>
    <Col
      className="mx-auto"
      md={4}
      sm={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Label
        className="bg-warning"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Note:
      </Label>
      <br />
      <textarea
        className="w-100"
        name="ef_note"
        onChange={[Function]}
        placeholder="Add a note for your expense (optional)"
        rows="5"
        value=""
      />
    </Col>
    <Col
      className="mx-auto"
      md={4}
      sm={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Button
        block={true}
        color="success"
        outline={true}
        tag="button"
      >
        Add_Expense
      </Button>
    </Col>
  </form>
</div>
`;

exports[`should render expense form correctly with one expense data 1`] = `
<div>
  <form
    name="expenseForm"
    noValidate={true}
    onSubmit={[Function]}
  >
    <Col
      className="mx-auto"
      md={4}
      sm={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Label
        className="bg-warning"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Description:
      </Label>
      <input
        autoFocus={true}
        className="w-100 form-control rounded-0"
        name="ef_description"
        onChange={[Function]}
        placeholder="Description..."
        type="text"
        value="Rent"
      />
    </Col>
    <Col
      className="mx-auto"
      md={4}
      sm={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Label
        className="bg-warning"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Amount:
      </Label>
      <input
        className="w-100 form-control rounded-0"
        name="ef_Amount"
        onChange={[Function]}
        placeholder="Amount..."
        type="text"
        value="1095"
      />
    </Col>
    <Col
      className="mx-auto"
      md={4}
      sm={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Badge
        color="warning"
        pill={true}
        tag="span"
      >
        Create Date:
      </Badge>
      <br />
      <withStyles(SingleDatePicker)
        anchorDirection="left"
        appendToBody={false}
        block={false}
        calendarInfoPosition="bottom"
        customCloseIcon={null}
        customInputIcon={null}
        date={"1969-12-27T23:00:00.000Z"}
        daySize={39}
        disableScroll={false}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        horizontalMonthPadding={13}
        id="sd-pIcker"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepFocusOnInput={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        noBorder={false}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsSelected": [Function],
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month.",
            "jumpToPrevMonth": "Move backward to switch to the previous month.",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
            "moveFocusByOneMonth": "Switch months.",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
            "openThisPanel": "Open this panel.",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field.",
            "selectFocusedDate": "Select the date in focus.",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        regular={false}
        renderCalendarInfo={null}
        renderDayContents={null}
        renderMonthElement={null}
        renderMonthText={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        small={false}
        verticalHeight={null}
        verticalSpacing={22}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </Col>
    <Col
      className="mx-auto"
      md={4}
      sm={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Label
        className="bg-warning"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Note:
      </Label>
      <br />
      <textarea
        className="w-100"
        name="ef_note"
        onChange={[Function]}
        placeholder="Add a note for your expense (optional)"
        rows="5"
        value="This is rent"
      />
    </Col>
    <Col
      className="mx-auto"
      md={4}
      sm={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Button
        block={true}
        color="success"
        outline={true}
        tag="button"
      >
        Add_Expense
      </Button>
    </Col>
  </form>
</div>
`;

exports[`should render expense form correctly. 1`] = `
<div>
  <form
    name="expenseForm"
    noValidate={true}
    onSubmit={[Function]}
  >
    <Col
      className="mx-auto"
      md={4}
      sm={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Label
        className="bg-warning"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Description:
      </Label>
      <input
        autoFocus={true}
        className="w-100 form-control rounded-0"
        name="ef_description"
        onChange={[Function]}
        placeholder="Description..."
        type="text"
        value=""
      />
    </Col>
    <Col
      className="mx-auto"
      md={4}
      sm={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Label
        className="bg-warning"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Amount:
      </Label>
      <input
        className="w-100 form-control rounded-0"
        name="ef_Amount"
        onChange={[Function]}
        placeholder="Amount..."
        type="text"
        value=""
      />
    </Col>
    <Col
      className="mx-auto"
      md={4}
      sm={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Badge
        color="warning"
        pill={true}
        tag="span"
      >
        Create Date:
      </Badge>
      <br />
      <withStyles(SingleDatePicker)
        anchorDirection="left"
        appendToBody={false}
        block={false}
        calendarInfoPosition="bottom"
        customCloseIcon={null}
        customInputIcon={null}
        date={"1970-01-01T00:00:00.000Z"}
        daySize={39}
        disableScroll={false}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        horizontalMonthPadding={13}
        id="sd-pIcker"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepFocusOnInput={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        noBorder={false}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsSelected": [Function],
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month.",
            "jumpToPrevMonth": "Move backward to switch to the previous month.",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
            "moveFocusByOneMonth": "Switch months.",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
            "openThisPanel": "Open this panel.",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field.",
            "selectFocusedDate": "Select the date in focus.",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        regular={false}
        renderCalendarInfo={null}
        renderDayContents={null}
        renderMonthElement={null}
        renderMonthText={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        small={false}
        verticalHeight={null}
        verticalSpacing={22}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </Col>
    <Col
      className="mx-auto"
      md={4}
      sm={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Label
        className="bg-warning"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Note:
      </Label>
      <br />
      <textarea
        className="w-100"
        name="ef_note"
        onChange={[Function]}
        placeholder="Add a note for your expense (optional)"
        rows="5"
        value=""
      />
    </Col>
    <Col
      className="mx-auto"
      md={4}
      sm={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Button
        block={true}
        color="success"
        outline={true}
        tag="button"
      >
        Add_Expense
      </Button>
    </Col>
  </form>
</div>
`;
